### PR TITLE
First draft of cancellation doc

### PR DESCRIPTION
**Pull request by gnossen**

### CODE REVIEW

**Comment by gnossen** [comment]
@temawi

File: content/en/docs/guides/cancellation.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
+---
+title: Cancellation
+description: >-
+  Explains how and when to cancel to RPCs.
```
Comment (sanjaypujare) [inline content/en/docs/guides/cancellation.md:4]: "to cancel RPCs"

File: content/en/docs/guides/cancellation.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+---
+
+### Overview
+
+When a gRPC client is no longer interested in the result of an RPC call, it may
+_cancel_ to signal this discontinuation of interest to the server.  After that
+point, the server should stop any ongoing computation and end its side of the
+connection. Often, servers are also clients to upstream servers, so that
```
Comment (sanjaypujare) [inline content/en/docs/guides/cancellation.md:12]: "connection" ? People might confuse this with the underlying TCP or HTTP connection

File: content/en/docs/guides/cancellation.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+A(Client) -->|Cancelled| B(Server 1)
+B-->|Cancelled|C(Server 2)
+```
+
+### Cancelling an RPC Call on the Client Side
+
+A client cancels an RPC call by calling a method on the call object or, in some
+languages, on the accompanying context object. At the level of the wire
+protocol, a client cancels an RPC call by sending an
+[`RST_STREAM`](https://datatracker.ietf.org/doc/html/rfc7540#section-6.4) frame
```
Comment (sergiitk) [inline content/en/docs/guides/cancellation.md:43]: RFC 7540 obsoleted by RFC 9113:  https://www.rfc-editor.org/rfc/rfc9113.html

File: content/en/docs/guides/cancellation.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+which will be called when a cancellation is received. This callback is expected
+to coordinate with the handler thread/coroutine to stop local processing on
+behalf of the request and clean up any associated resources, including upstream
+RPCs. In other languages, the mechanism may be a multithreading Event object,
+which will be set when the RPC has been cancelled.  Regardless of the
+language-specific mechanism, the thread/coroutine processing the RPC must
+periodically check if RPC has been cancelled and if it has, cease processing and
+begin the clean-up process.
+
+```mermaid
```
Comment (XuanWang-Amos) [inline content/en/docs/guides/cancellation.md:67]: Can we add a description for this graph?

File: content/en/docs/guides/cancellation.md (hunk 3 above)
Comment (sergiitk) [reply content/en/docs/guides/cancellation.md:43]: Oh, it also comes with nicer the anchors:  https://www.rfc-editor.org/rfc/rfc9113#CANCEL

File: content/en/docs/guides/cancellation.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+```
+
+### Cancelling an RPC Call on the Client Side
+
+A client cancels an RPC call by calling a method on the call object or, in some
+languages, on the accompanying context object. At the level of the wire
+protocol, a client cancels an RPC call by sending an
+[`RST_STREAM`](https://datatracker.ietf.org/doc/html/rfc7540#section-6.4) frame
+with an error code of
+[`CANCEL(8)`](https://datatracker.ietf.org/doc/html/rfc7540#section-7). gRPC
```
Comment (sanjaypujare) [inline content/en/docs/guides/cancellation.md:45]: Do we need to provide this level of detail in a user doc? A gRFC or a technical doc yes but why a user doc?

File: content/en/docs/guides/cancellation.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+clients do not provide additional details to the server about the reason for the
+cancellation. When the server receives this frame, the application-provided
+server handler may be busy processing the request in a thread, coroutine, or
+other unit of concurrency. The gRPC library in general does not have a mechanism
+to interrupt the server handler, so the server handler must coordinate with the
+gRPC library to ensure that local processing of the request ceases and any
+upstream RPC calls initiated in service of this RPC call are themselves
+cancelled.
+
```
Comment (larry-safran) [inline content/en/docs/guides/cancellation.md:54]: Talk about the cancel call taking a description (even though it isn't passed to the server) which can be used for logging or throwing exceptions.

**Review by ejona86 — COMMENTED** [review]
Deadline triggers cancellation when it fires. We should cross-mention that in the docs as appropriate.

Cancellation occurs due to I/O errors as well.

File: content/en/docs/guides/cancellation.md [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+languages, on the accompanying context object. At the level of the wire
+protocol, a client cancels an RPC call by sending an
+[`RST_STREAM`](https://datatracker.ietf.org/doc/html/rfc7540#section-6.4) frame
+with an error code of
+[`CANCEL(8)`](https://datatracker.ietf.org/doc/html/rfc7540#section-7). gRPC
+clients do not provide additional details to the server about the reason for the
+cancellation. When the server receives this frame, the application-provided
+server handler may be busy processing the request in a thread, coroutine, or
+other unit of concurrency. The gRPC library in general does not have a mechanism
+to interrupt the server handler, so the server handler must coordinate with the
```
Comment (sanjaypujare) [inline content/en/docs/guides/cancellation.md:50]: I think what is implied here is that the application code needs to do this. "server handler" as application code may not be obvious

File: content/en/docs/guides/cancellation.md [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+gRPC library to ensure that local processing of the request ceases and any
+upstream RPC calls initiated in service of this RPC call are themselves
+cancelled.
+
+The exact mechanism for notifying the server handler that the RPC being handled
```
Comment (sergiitk) [inline content/en/docs/guides/cancellation.md:55]: Asked bard to shrink :)

> The mechanism for notifying the server handler of an RPC cancellation differs between languages. Some languages use a callback function, while others use an event object. Regardless of the mechanism, the thread/coroutine processing the RPC must periodically check if it has been cancelled and, if so, cease processing and begin cleanup.

File: content/en/docs/guides/cancellation.md [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    subgraph Thread1
+      direction TB
+      grpc("gRPC library") -->|calls| cb("onCancel CB")
+      cb --> set("sets cancelled to true")
+    end
+    subgraph Thread2
+      direction TB
+      cancelled{cancelled?} -->|false| perform("perform some work")
+      perform --> cancelled
+      cancelled -->|true| cleanup("cancel upstream RPCs")
```
Comment (sanjaypujare) [inline content/en/docs/guides/cancellation.md:82]: And also do other clean up like clean up resources (close file handles etc)

File: content/en/docs/guides/cancellation.md [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+  subgraph Server2
+  end
+
+  Client -->|CANCEL| Server1
+  Server1 -->|CANCEL| Server2
+```
+
+### Language Support
+
+| Language | Example        | Notes                            |
```
Comment (sanjaypujare) [inline content/en/docs/guides/cancellation.md:96]: Notes column is empty - nothing to add there?

File: content/en/docs/guides/cancellation.md (hunk 3 above)
Comment (gnossen) [reply content/en/docs/guides/cancellation.md]: Thanks for the suggestion. This section was removed for being too technical.

File: content/en/docs/guides/cancellation.md (hunk 4 above)
Comment (gnossen) [reply content/en/docs/guides/cancellation.md:47]: The graph has been simplified to gloss over this. The answer to that question varies from language to language.

File: content/en/docs/guides/cancellation.md (hunk 5 above)
Comment (gnossen) [reply content/en/docs/guides/cancellation.md]: Removed.

File: content/en/docs/guides/cancellation.md (hunk 7 above)
Comment (gnossen) [reply content/en/docs/guides/cancellation.md]: More instances of "application-provided" have been added.

File: content/en/docs/guides/cancellation.md (hunk 8 above)
Comment (gnossen) [reply content/en/docs/guides/cancellation.md]: Shrunk.

File: content/en/docs/guides/cancellation.md (hunk 10 above)
Comment (gnossen) [reply content/en/docs/guides/cancellation.md:67]: Notes added

**Review by temawi — APPROVED** [review]

**Pull request merged**